nbtc/relayer.go

```diff
@@ -0,0 +1,165 @@
+package nbtc
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/rpcclient"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/gonative-cc/relayer/bitcoin"
+	"github.com/gonative-cc/relayer/dal"
+	"github.com/rs/zerolog/log"
+)
+
+// Relayer broadcasts pending transactions from the database to the Bitcoin network.
+type Relayer struct {
+	db           *dal.DB
+	btcClient    bitcoin.Client
+	shutdownChan chan struct{}
+}
+
+// NewRelayer creates a new Relayer instance with the given configuration.
+func NewRelayer(db *dal.DB) (*Relayer, error) {
+	if db == nil {
+		return nil, fmt.Errorf("database cannot be nil")
+	}
+
+	host := os.Getenv("BTC_RPC")
+	user := os.Getenv("BTC_RPC_USER")
+	pass := os.Getenv("BTC_RPC_PASS")
+
+	if host == "" || user == "" || pass == "" {
+		err := fmt.Errorf("missing env variables with Bitcoin node configuration")
+		log.Err(err).Msg("")
+		return nil, err
+	}
+
+	connCfg := &rpcclient.ConnConfig{
+		Host:         host,
+		User:         user,
+		Pass:         pass,
+		HTTPPostMode: true,
+		DisableTLS:   false,
+	}
+
+	client, err := rpcclient.New(connCfg, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Relayer{
+		db:           db,
+		btcClient:    client,
+		shutdownChan: make(chan struct{}),
+	}, nil
+}
+
+// Start starts the relayer's main loop to broadcast transactions.
+func (r *Relayer) Start() error {
+	ticker := time.NewTicker(time.Second * 2)
+	ticker2 := time.NewTicker(time.Second * 4) //TODO: this probably can run every minute or even more
+
+	for {
+		select {
+		case <-r.shutdownChan:
+			r.btcClient.Shutdown()
+			return nil
+		case <-ticker.C:
+			if err := r.processPendingTxs(); err != nil {
+				//TODO: add proper error handling here and decide which errors we shutdown the relayer
+				if err.Error() == "error updating transaction status" {
+					log.Err(err).Msg("Critical error updating transaction status, shutting down")
+					close(r.shutdownChan)
+				} else {
+					log.Err(err).Msg("Error processing transactions, continuing")
+				}
+			}
+		case <-ticker2.C:
+			r.checkConfirmations()
```

Comment:
@robert-zaremba @vuvoth Here both the `processPendingTxs` and `checkConfirmations` will be operating on the same database. Is there a risk for a race con? Should use mutex on the db? Please let me know what you guys think